system/lib/wasmfs/backend.h

```diff
@@ -0,0 +1,31 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the modular backend abstract class.
+// Other file system backends can use this to interface with the new file
+// system. Current Status: Work in Progress. See
+// https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include "file.h"
+#include <memory>
+#include <stdlib.h>
+#include <utility>
+#include <wasi/api.h>
+
+#pragma once
+
+namespace wasmfs {
+class Backend {
+public:
+  virtual std::shared_ptr<DataFile> createFile(mode_t mode) = 0;
+  virtual std::shared_ptr<Directory> createDirectory(mode_t mode) = 0;
+  virtual ~Backend() = default;
+};
+
+// This will return an instance of a MemoryFileBackend.
+// Note: Backends will be defined in cpp files, but functions to acquire them
```

Comment:
Do you mean that it will no longer be possible to mount and implement file systems purely in JS?

Reply:
We will provide the ability to store files in JS. However, the file structure itself will be in C++ to take advantage of multi-threading. 

So a file system where files are stored in JS is still an option.

Reply:
We can also add a JS API to call into C++ to do mounting etc. The API might end up looking a little different though, that's not clear yet I think.

Reply:
Just as long as there are clear examples (and ideally docs) for how to write custom FSes in the new design. Though I guess I'll have to learn a little C++ - so far I've been able to avoid that!